csrc/preseg_passes/move_split_cat.cpp

```diff
@@ -58,79 +130,90 @@ class CancelSplitCat {
 
   Fusion* fusion_;
 
+  // `id_model_` is supposed to be read-only and reflect the
+  // original fusion.
   IdModel id_model_;
+
+  // `id_model_for_merging_` is used for
+  // `sameIterDomainTransforms`, which unionizes IterDomains in slices and
+  // checks IterDomains in pads are unionized in the same way.
+  IdModel id_model_for_merging_;
 };
 
-bool CancelSplitCat::horizontallyMergeable(
-    const std::vector<Expr*>& frontier,
-    int64_t& split_axis) {
-  NVF_ERROR(!frontier.empty());
-
-  // Check all Exprs in `frontier`
-  // 1. have the same op type and attributes,
-  // 2. transform IDs in the same way, and
-  // 3. don't resize the split axis.
-
-  if (std::adjacent_find(
-          frontier.begin(), frontier.end(), [](Expr* lhs, Expr* rhs) {
-            return !lhs->sameOp(rhs);
-          }) != frontier.end()) {
-    return false;
-  }
+bool sameOp(const std::vector<Expr*>& frontier) {
+  return std::adjacent_find(
+             frontier.begin(), frontier.end(), [](Expr* lhs, Expr* rhs) {
+               return !lhs->sameOp(rhs);
+             }) == frontier.end();
+}
 
-  if (auto* set = dynamic_cast<LoadStoreOp*>(frontier[0])) {
-    if (set->opType() == LoadStoreOpType::Set) {
-      auto* set_out = set->out()->as<TensorView>();
-      std::optional<std::vector<int64_t>> permutation =
-          ir_utils::computePermutation(
-              set_out->getRootDomain(), set_out->getMaybeRFactorDomain());
-      if (!permutation.has_value()) {
+bool CancelSplitCat::sameIterDomainTransforms(
+    const std::vector<SliceOp*>& slices,
+    const std::vector<PadOp*>& pads) {
+  NVF_ERROR(slices.size() == pads.size());
+  NVF_ERROR(!slices.empty());
+
+  ValGraph& exact_graph = id_model_for_merging_.idGraph(IdMappingMode::EXACT);
+  {
+    const std::vector<IterDomain*>& first_rfactor =
+        slices[0]->out()->getMaybeRFactorDomain();
+    size_t num_dims = first_rfactor.size();
+    for (size_t i = 1; i < slices.size(); i++) {
+      const std::vector<IterDomain*>& rfactor =
+          slices[i]->out()->getMaybeRFactorDomain();
+      if (rfactor.size() != num_dims) {
         return false;
       }
+      for (size_t j = 0; j < num_dims; j++) {
+        exact_graph.mapVals(first_rfactor[j], rfactor[j]);
+      }
+    }
+  }
 
-      for (size_t i = 1; i < frontier.size(); i++) {
-        auto* other_set_out =
-            frontier[i]->as<LoadStoreOp>()->out()->as<TensorView>();
-        std::optional<std::vector<int64_t>> other_permutation =
-            ir_utils::computePermutation(
-                other_set_out->getRootDomain(),
-                other_set_out->getMaybeRFactorDomain());
-        if (!other_permutation.has_value()) {
-          return false;
-        }
-        if (*permutation != *other_permutation) {
+  for (PadOp* pad : pads) {
+    auto* pad_out = pad->out()->as<TensorView>();
+    if (id_model_for_merging_.hasSelfMapping(pad_out)) {
+      return false;
```

Comment:
I figure this is just a sanity check here? in case there's some downstream logic that might render our previous manual mapping invalid?
I think this is worth a comment.

Reply:
@wujingyue Any particular example?

Reply:
Added a comment with an example link.